phd-tests/framework/src/test_vm/mod.rs

```diff
@@ -868,7 +913,21 @@ impl TestVm {
     /// waits for another shell prompt to appear using
     /// [`Self::wait_for_serial_output`] and returns any text that was buffered
     /// to the serial console after the command was sent.
-    pub async fn run_shell_command(&self, cmd: &str) -> Result<String> {
+    ///
+    /// After running the shell command, sends `echo $?` to query and return the
+    /// command's return status as well.
+    // TO REVIEWERS: it would be really nice to write this as a function
+    // returning a `struct ShellCommandExecutor` that impls
+    // `Future<Output=String>` where the underlying ShellOutput is automatically
+    // `.expect_ok()`'d. In such a case it would be possible for the struct to
+    // have an `.expect_err()` that replaces teh default `.expect_ok()`
+    // behavior, so that the likely case doesn't need any change in PHD tests.
+    //
+    // unfortunately I don't know how to plumb the futures for that, since we'd
+    // have to close over `&self`, so doing any Boxing to hold an
+    // `async move {}` immediately causes issues.
```

Comment:
my thinking is that it'd be nice if `vm.run_shell_command("foo").await` automatically did what `expect_ok` does in this change, but if you know to expect an error you could `vm.run_shell_command("foo").expect_err().await` to modify the future and _then_ run it. concretely it looked something like:

```rust
struct ShellCommandExecutor {
    fut: Pin<Box<impl Future<Output=Result<ShellOutput>> + Send + Sync>>
}

impl ShellCommandExecutor {
    fn expect_err(self) -> ShellCommandExecutorExpectErr { /* build the thing */ }
}
impl Future for ShellCommandExecutor {
    fn poll(&self, cx: &mut Context) -> Self::Output {
        match self.fut.poll(cx) {
            Poll::Ready(v) => Poll::Ready(v.expect_ok()),
            Poll::Pending => Poll::Pending
        }
    }
}
```

but i couldn't figure out how to `impl Future` in a way that doesn't conflict with `fut` holding `&self`. so this patch does the next best thing i could think of.

Reply:
This feels like it's what [`IntoFuture`](https://doc.rust-lang.org/stable/std/future/trait.IntoFuture.html) is supposed to be for, right? "I want to have a thing which is not actually a `Future` implementation, but knows how to turn itself into a future, so you can still `.await` it and it will go and turn itself into a future".

Reply:
ok! now that i got the more Future-ful approach together i think this all is _much_ cleaner (and incidentally resolves your other comments). it's kinda sickrad that this only comes with like 10 changes in tests now, since it's going along with the general assumption that commands run in guests _should_ be successful.